Possible crash in putenv

> In file included from test\test_system.cpp:11:0:
> ..\../boost/nowide/cenv.hpp: In function 'int boost::nowide::putenv(char*)':
> ..\../boost/nowide/cenv.hpp:104:41: warning: comparison between pointer and
> zero character constant [-Wpointer-compare]
>         while(*key_end!='=' && key_end!='\0')

--- a/test/test_env.cpp
+++ b/test/test_env.cpp
@@ -35,6 +35,9 @@ int main()
         TEST(boost::nowide::getenv("BOOST_TEST2"));
         TEST(boost::nowide::getenv("BOOST_TEST_INVALID")==0);
         TEST(boost::nowide::getenv("BOOST_TEST2")==example + "x");
+#ifdef BOOST_WINDOWS
+        // Passing a variable without an equals sign (before \0) is an error
+        // But GLIBC has an extension that unsets the env var instead
         char penv2[256] = {0};
         const char* sPenv2 = "BOOST_TEST1SOMEGARBAGE=";
         strncpy(penv2, sPenv2, sizeof(penv2) - 1);
@@ -43,6 +46,7 @@ int main()
         TEST(boost::nowide::putenv(penv2)==-1);
         TEST(boost::nowide::getenv("BOOST_TEST1"));
         TEST(boost::nowide::getenv("BOOST_TEST1")==example);
+#endif
     }
     catch(std::exception const &e) {
         std::cerr << "Failed " << e.what() << std::endl;

--- a/test/test_env.cpp
+++ b/test/test_env.cpp
@@ -35,6 +35,14 @@ int main()
         TEST(boost::nowide::getenv("BOOST_TEST2"));
         TEST(boost::nowide::getenv("BOOST_TEST_INVALID")==0);
         TEST(boost::nowide::getenv("BOOST_TEST2")==example + "x");
+        char penv2[256] = {0};
+        const char* sPenv2 = "BOOST_TEST1SOMEGARBAGE=";
+        strncpy(penv2, sPenv2, sizeof(penv2) - 1);
+        // End the string before the equals sign -> Expect fail
+        penv2[strlen("BOOST_TEST1")] = '\0';
+        TEST(boost::nowide::putenv(penv2)==-1);
+        TEST(boost::nowide::getenv("BOOST_TEST1"));
+        TEST(boost::nowide::getenv("BOOST_TEST1")==example);
     }
     catch(std::exception const &e) {
         std::cerr << "Failed " << e.what() << std::endl;

